tools/clang/lib/Sema/SemaDXR.cpp

```diff
@@ -1254,12 +1258,165 @@ void DiagnoseClosestHitEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
+         .Case("RWDispatchNodeInputRecord",
+               launchType == DXIL::NodeLaunchType::Broadcasting)
+         .Case("GroupNodeInputRecords",
+               launchType == DXIL::NodeLaunchType::Coalescing)
+         .Case("RWGroupNodeInputRecords",
+               launchType == DXIL::NodeLaunchType::Coalescing)
+         .Case("EmptyNodeInput",
+               launchType == DXIL::NodeLaunchType::Coalescing)
+         .Case("ThreadNodeInputRecord",
+               launchType == DXIL::NodeLaunchType::Thread)
+         .Case("RWThreadNodeInputRecord",
+               launchType == DXIL::NodeLaunchType::Thread)
+         .Default(false);
+}
+
 void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
+
+  SourceLocation computeLoc = SourceLocation();
+  SourceLocation nodeLoc = SourceLocation();
+  SourceLocation nodeLaunchLoc = SourceLocation();
+  DXIL::NodeLaunchType nodeLaunchType = DXIL::NodeLaunchType::Invalid;
+  unsigned inputCount = 0;
+
+  // a function may be both compute and work-graph node
+  for (auto *pAttr : FD->specific_attrs<HLSLShaderAttr>()) {
+    DXIL::ShaderKind shaderKind = ShaderModel::KindFromFullName(pAttr->getStage());
+    if (shaderKind == DXIL::ShaderKind::Node) {
+      nodeLoc = pAttr->getLocation();
+    } else if (shaderKind == DXIL::ShaderKind::Compute) {
+      computeLoc = pAttr->getLocation();
+    }
+  }
+  // if this isn't a work-graph node we can quit now
+  if (!nodeLoc.isValid())
+    return;
+
+  // save NodeLaunch type for use later
+  if (auto nodeLaunchAttr = FD->getAttr<HLSLNodeLaunchAttr>()) {
+    nodeLaunchType =
+      ShaderModel::NodeLaunchTypeFromName(nodeLaunchAttr->getLaunchType());
+    nodeLaunchLoc = nodeLaunchAttr->getLocation();
+  } else {
+    nodeLaunchType = DXIL::NodeLaunchType::Broadcasting;
+    nodeLaunchLoc = SourceLocation();
+  }
+
+  // If this is both a compute shader and work-graph node, it may only have
+  // broadcasting launch mode
+  if (computeLoc.isValid() &&
+      nodeLaunchType != DXIL::NodeLaunchType::Broadcasting) {
+    S.Diags.Report(nodeLaunchLoc, diag::err_hlsl_compute_launch_compatibility)
+      << FD->getName() << ShaderModel::GetNodeLaunchTypeName(nodeLaunchType);
+    S.Diags.Report(computeLoc, diag::note_defined_here) << "compute";
+  }
+
+  // Check that if a Thread launch node has the NumThreads attribute the
+  // thread group size is (1,1,1)
+  if (nodeLaunchType == DXIL::NodeLaunchType::Thread) {
+    if (auto NumThreads = FD->getAttr<HLSLNumThreadsAttr>()) {
+      if (NumThreads->getX() != 1 || NumThreads->getY() != 1 ||
+          NumThreads->getZ() != 1) {
+        S.Diags.Report(NumThreads->getLocation(),
+                       diag::err_hlsl_wg_thread_launch_group_size)
+          << NumThreads->getRange();
+        // Only output the note if the source location is valid
+        if (nodeLaunchLoc.isValid())
+          S.Diags.Report(nodeLaunchLoc, diag::note_defined_here)
+              << "Launch type";
```

Comment:
I'm not sure a note indicating where the launch type is defined is necessary. There are several other errors here that reference the launch type, but don't include this note. I think it's clear enough where to find it without the note.

Reply:
Then I'm afraid there are a fair number of other errors that might similarly benefit